data/sqlutil/macros.go

```diff
@@ -189,53 +193,67 @@ func getMacroMatches(input string, name string) ([][]string, error) {
 			macroEnd = matched[matchedIndex][1] - macroStart
 		}
 		macroString := inputCopy[0:macroEnd]
-		macroMatch := []string{macroString}
+		macroMatch := Macro{Name: macroString}
 
 		args := ""
 		// if opening parenthesis was found, extract contents as arguments
 		if argStart > 0 {
 			args = inputCopy[argStart : macroEnd-1]
 		}
-		macroMatch = append(macroMatch, args)
+		macroMatch.Args = parseArgs(args)
 		matchedMacros = append(matchedMacros, macroMatch)
 	}
 	return matchedMacros, nil
 }
 
-// Interpolate returns an interpolated query string given a backend.DataQuery
-func Interpolate(query *Query, macros Macros) (string, error) {
-	for key, defaultMacro := range DefaultMacros {
-		if _, ok := macros[key]; !ok {
-			// If the driver doesn't define some macro, use the default one
-			macros[key] = defaultMacro
+func parseArgs(args string) []string {
+	argsArray := []string{}
+	phrase := []rune{}
+	bracketCount := 0
+	for _, v := range args {
+		phrase = append(phrase, v)
+		if v == '(' {
+			bracketCount++
+			continue
+		}
+		if v == ')' {
+			bracketCount--
+			continue
+		}
+		if v == ',' && bracketCount == 0 {
+			removeComma := phrase[:len(phrase)-1]
+			argsArray = append(argsArray, string(removeComma))
+			phrase = []rune{}
 		}
 	}
+	argsArray = append(argsArray, strings.TrimSpace(string(phrase)))
+	return argsArray
+}
+
+// Interpolate returns an interpolated query string given a backend.DataQuery
+func Interpolate(query *Query, macros Macros) (string, error) {
+	mergedMacros := Macros{}
+	maps.Copy(mergedMacros, DefaultMacros)
+	maps.Copy(mergedMacros, macros)
+
 	rawSQL := query.RawSQL
 
-	for key, macro := range macros {
+	for key, macro := range mergedMacros {
 		matches, err := getMacroMatches(rawSQL, key)
 		if err != nil {
 			return rawSQL, err
 		}
+		if len(matches) == 0 {
+			continue
+		}
 
 		for _, match := range matches {
-			if len(match) == 0 {
-				// There were no matches for this macro
-				continue
-			}
-
-			args := []string{}
-			if len(match) > 1 {
-				// This macro has arguments
-				args = trimAll(strings.Split(match[1], ","))
-			}
-
-			res, err := macro(query.WithSQL(rawSQL), args)
+			res, err := macro(query.WithSQL(rawSQL), match.Args)
 			if err != nil {
 				return rawSQL, err
 			}
 
-			rawSQL = strings.ReplaceAll(rawSQL, match[0], res)
+			rawSQL = strings.ReplaceAll(rawSQL, match.Name, res)
```

Comment:
random note, this is not part of the diff.. here we use `ReplaceAll()`.
in `sqlds`, it's `Replace()`:

https://github.com/grafana/sqlds/blob/dda2dc0a54b128961fc9f7885baabf555f3ddfdc/macros.go#L258

what do you think, which one is "more correct"?

Reply:
Ah, I actually made the change intentionally in my original PR; this time around I forgot, but then the linter in this repo told me to change it! Looks like this comes from [gocritic](https://github.com/go-critic/go-critic/issues/872):
```
❯ mage lint
data/sqlutil/macros.go:215:13: wrapperFunc: use strings.ReplaceAll method in `strings.Replace(rawSQL, match.full, res, -1)` (gocritic)
			rawSQL = strings.Replace(rawSQL, match.full, res, -1)
			         ^
Error: running "golangci-lint run ./..." failed with exit code 1
```

I also prefer `ReplaceAll` as I think it makes the intention more clear, but it's a pretty minor thing.